File: MattEland.Emergence.Domain/Position.fs
namespace MattEland.Emergence.Domain

open System

type MoveDirection =
  | Up = 0
  | Right = 1
  | Down = 2
  | Left = 3

[<CustomEquality>]
[<NoComparison>]
type Position (x:int, y:int) =
  struct
    member this.X = x
    member this.Y = y
    
    member this.Add (pos: Position) : Position = new Position(x + pos.X, y + pos.Y)
    member this.Subtract (pos: Position) : Position = new Position(x - pos.X, y - pos.Y)

    override this.ToString() = "{" + x.ToString() + ", " + y.ToString() + "}"

    override this.GetHashCode() = (x.GetHashCode() * 100) + y.GetHashCode()
      
    override this.Equals(b) =
      match b with
      | :? Position as p -> p.X = x && p.Y = y
      | _ -> false

    member this.GetNeighbor(direction: MoveDirection): Position =
      match direction with
      | MoveDirection.Up -> new Position(x, y - 1)
      | MoveDirection.Right -> new Position(x + 1, y)
      | MoveDirection.Down -> new Position(x, y + 1)
      | MoveDirection.Left -> new Position(x - 1, y)
      | _ -> raise (NotSupportedException("The direction " + direction.ToString() + " is not supported for position modification"))
  end

File: MattEland.Emergence.Model/Position.cs
using System;

namespace MattEland.Emergence.Model
{
    public struct Position : IEquatable<Position>
    {
        public Position(int x, int y)
        {
            X = x;
            Y = y;
        }

        public int X { get; }
        public int Y { get; }

        public Position Add(Position pos) => new Position(X + pos.X, Y + pos.Y);
        public Position Add(int deltaX, int deltaY) => new Position(X + deltaX, Y + deltaY);
        public Position Subtract(Position pos) => new Position(X - pos.X, Y - pos.Y);
        public Position Subtract(int deltaX, int deltaY) => new Position(X - deltaX, Y - deltaY);

        public bool Equals(Position other)
        {
            return X == other.X && Y == other.Y;
        }

        public override bool Equals(object obj)
        {
            return obj is Position other && Equals(other);
        }

        public override int GetHashCode()
        {
            unchecked
            {
                return (X * 397) ^ Y;
            }
        }

        public static bool operator ==(Position left, Position right)
        {
            return left.Equals(right);
        }

        public static bool operator !=(Position left, Position right)
        {
            return !left.Equals(right);
        }

        public override string ToString() => $"({X}, {Y})";

        public Position GetNeighbor(MoveDirection direction)
        {
            switch (direction)
            {
                case MoveDirection.Up:
                    return Add(0, -1);
                case MoveDirection.Right:
                    return Add(1, 0);
                case MoveDirection.Down:
                    return Add(0, 1);
                case MoveDirection.Left:
                    return Add(-1, 0);
                default:
                    throw new ArgumentOutOfRangeException(nameof(direction), direction, null);
            }
        }
    }
}
